File: backend/app/models/farmer.py
from sqlalchemy import Column, String, Float, ForeignKey, DateTime, Text, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
from app.database import Base


class Farmer(Base):
    __tablename__ = "farmers"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id"), unique=True, nullable=False)
    name = Column(String(255), nullable=False)
    district = Column(String(100), nullable=False)
    state = Column(String(100), nullable=False, default="Tamil Nadu")
    soil_health_card_id = Column(String(50), nullable=True)
    agristack_id = Column(String(50), nullable=True)
    total_land_acres = Column(Float, default=0.0)
    profile_data = Column(JSON, default={})
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="farmer")
    farms = relationship("Farm", back_populates="farmer", cascade="all, delete-orphan")

File: mobile/lib/models/farmer.dart
class Farmer {
  final String id;
  final String name;
  final String district;
  final String state;
  final double totalLandAcres;
  final String? soilHealthCardId;

  const Farmer({
    required this.id,
    required this.name,
    required this.district,
    required this.state,
    this.totalLandAcres = 0.0,
    this.soilHealthCardId,
  });

  factory Farmer.fromJson(Map<String, dynamic> json) {
    return Farmer(
      id: json['id']?.toString() ?? '',
      name: json['name'] as String? ?? '',
      district: json['district'] as String? ?? '',
      state: json['state'] as String? ?? '',
      totalLandAcres: (json['total_land_acres'] as num?)?.toDouble() ?? 0.0,
      soilHealthCardId: json['soil_health_card_id'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'district': district,
        'state': state,
        'total_land_acres': totalLandAcres,
        'soil_health_card_id': soilHealthCardId,
      };
}
